rabbitmq_integration_test.go

```diff
@@ -318,51 +312,81 @@ func TestRabbitMqClient_PublishWithContext(t *testing.T) {
 }
 
 func TestRabbitMqClient_Close(t *testing.T) {
-	t.Run("ShouldRetunNoError_WhenIsClosedTrue", func(t *testing.T) {
+	t.Run("ShouldRetunNoError_WhenCloseRabbitMq", func(t *testing.T) {
 		// Init
 		rabbitClient, errClient := NewRabbitMqClient(os.Getenv("RABBITMQ_URL"), WithConsumerThread(1))
 		require.NoError(t, errClient)
 
-		//WhenIsClosedTrue
-		rabbitClient.isClosed.Store(true)
-
 		// Code under test
 		err := rabbitClient.Close()
-
 		// Expectations
 		// -- ShouldRetunNoError
 		require.NoError(t, err)
+
+		// Code under test
+		err = rabbitClient.Close()
+
+		// -- ShouldRetunNoError when call Close again
+		require.NoError(t, err)
+		require.Equal(t, rabbitClient.isClosed.Load(), true)
 	})
-	t.Run("ShouldCloseRabbitMq", func(t *testing.T) {
+
+	t.Run("ShouldCloseRabbitMqUntilConsumerFinish", func(t *testing.T) {
 		// Init
-		rabbitClient, errClient := NewRabbitMqClient(os.Getenv("RABBITMQ_URL"), WithConsumerThread(0))
+		rabbitClient, errClient := NewRabbitMqClient(os.Getenv("RABBITMQ_URL"), WithConsumerThread(1))
 		require.NoError(t, errClient)
+
 		testTopicName := "msgbuzz.pubtest"
 		actualMsgReceivedChan := make(chan []byte)
-		rabbitClient.isClosed.Store(false)
+		var consumerFinish string
 
 		// -- listen topic
 		rabbitClient.On(testTopicName, "msgbuzz", func(confirm MessageConfirm, bytes []byte) error {
+			defer func() {
+				require.NoError(t, confirm.Ack())
+			}()
+
+			// wait consumer finish
+			time.Sleep(2 * time.Second)
+
 			actualMsgReceivedChan <- bytes
-			return confirm.Ack()
+			consumerFinish = "consumer_finish"
+
+			// publish random topic
+			err := rabbitClient.Publish("msgbuzz.pubtest-random", []byte("Hi from msgbuzz"))
+			// -- ShouldRetunNoError
+			require.NoError(t, err)
+
+			return nil
 		})
 
 		go rabbitClient.StartConsuming()
-
 		// -- wait for exchange and queue to be created
 		time.Sleep(1 * time.Second)
 
+		//publish topic
+		sentMessage := []byte("some msg from msgbuzz")
+		err := rabbitClient.Publish(testTopicName, sentMessage)
+		require.NoError(t, err)
+
+		// -- Should receive correct msg
+		waitSec := 20
+		select {
+		case <-time.After(time.Duration(waitSec) * time.Second):
+			t.Fatalf("Not receiving msg after %d seconds", waitSec)
+		case actualMessageReceived := <-actualMsgReceivedChan:
+			require.Equal(t, sentMessage, actualMessageReceived)
+		}
+
 		// Code under test
-		err := rabbitClient.Close()
+		err = rabbitClient.Close()
+
 		// Expectations
 		// -- ShouldRetunNoError
 		require.NoError(t, err)
 		require.Equal(t, rabbitClient.isClosed.Load(), true)
 
-		//should return error no channel when try to publish after close()
-		err = rabbitClient.Publish(testTopicName, []byte("Hi from msgbuzz"))
-		require.Error(t, err)
-		require.Equal(t, strings.Contains(err.Error(), "error when getting channel"), true)
-
+		// assert variable for consumer finish
+		require.Equal(t, "consumer_finish", consumerFinish)
```

Comment:
@bintangsibarani I think we can remove `consumerFinish`. actualMessageRecieved is suffice IMO